### PR TITLE
ascanrules: Address External Redirect False Positives

### PR DESCRIPTION
**Pull request by kingthorin**
## Overview
The External Redirect scan rules has been updated to account for potential false positives involving JavaScript comments.

### CODE REVIEW

**Comment by psiinon** [comment]
Logo
**Checkmarx One – Scan Summary & Details** – [2b9816e5\-290f\-41f3\-89fa\-e369ec35e7d3](https://eu.ast.checkmarx.net/projects/0da811fa-6373-42b1-bbb6-d0026285b136/scans?id=2b9816e5-290f-41f3-89fa-e369ec35e7d3&branch=kingthorin%2Fredir-wavsep-fps)
<details><summary><b>Fixed Issues (1)</b></summary>
Great job! The following issues were fixed in this Pull Request

Severity|Issue|Source File / Package
---|---|---
LOW|~~[Heap_Inspection](https://eu.ast.checkmarx.net/results/0da811fa-6373-42b1-bbb6-d0026285b136/2b9816e5-290f-41f3-89fa-e369ec35e7d3/sast/description/244/8423996506624030647)~~|/addOns/authhelper/src/test/java/org/zaproxy/addon/authhelper/AuthUtilsUnitTest\.java: [1346](https://github.com/kingthorin/zap-extensions/blob/redir-wavsep-fps//addOns/authhelper/src/test/java/org/zaproxy/addon/authhelper/AuthUtilsUnitTest.java#L1346)
</details>



---
Use @Checkmarx to reach out to us for assistance.

Just send a PR comment with `@Checkmarx` followed by a natural language request.

**Examples:** `@Checkmarx how are you able to help me?` `@Checkmarx rescan this PR`

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java [hunk 1]
```diff
@@ -531,6 +531,68 @@ void shouldReportRedirectWithJsLocationMethods(String jsMethod) throws Exception
...
+                        "console.log(\"example\"); // console.log('@@@content@@@');"),
+                Arguments.of(
+                        "Inline single line (w/ unicode escape)",
+                        "console.log(\"🔥 example\"); // console.log('\u1F525 @@@content@@@');"),
+                // Next needs double escape because of Java
+                Arguments.of(
+                        "Inline single line (w/ hex escape)",
+                        "console.log(\"example\"); // console.log('\\xD83D @@@content@@@');"),
+                Arguments.of(
+                        "Inline single line (w/ java escapes)",
```
Comment (kingthorin) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: It occurred to me this should be single character escape not "Java"...

I'll tweak that in Monday but I'd still be happy for reviews.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java [hunk 2]
```diff
@@ -489,11 +492,176 @@ private static RedirectType isRedirected(String payload, HttpMessage msg) {
...
+                && JS_PRE_CHECKS.stream()
+                        .noneMatch(chk -> StringUtils.containsIgnoreCase(value, chk))) {
+            return false;
+        }
+        Set<String> extractedComments = extractJsComments(value);
+        String[] valueWithoutComments = {value};
+        extractedComments.forEach(
+                comment -> valueWithoutComments[0] = valueWithoutComments[0].replace(comment, ""));
+
+        Matcher matcher = pattern.matcher(valueWithoutComments[0]);
```
[the remark this answers is not in the document]
Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: @psiinon / @thc202 thoughts on this?

[the remark this answers is not in the document]
Comment (thc202) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: It's true that's an anti-pattern but the suggested change is not clear either. This should use an enhanced for instead IMO.

[the remark this answers is not in the document]
Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Thanks

**Comment by kingthorin** [comment]
Deconflicted

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java [hunk 3]
```diff
@@ -531,6 +531,68 @@ void shouldReportRedirectWithJsLocationMethods(String jsMethod) throws Exception
         assertThat(alertsRaised.get(0).getEvidence().startsWith(HttpHeader.HTTP), equalTo(true));
     }
 
+    private static Stream<Arguments> provideCommentStrings() {
```
Comment (thc202) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: - This should have malformed escapes and comments.
 - Better run the test coverage, no escapes are actually being tested (the escape char needs to be escaped).
 - All comments should lead to an alert if it wasn't for the fix, if you revert the scan rule changes and run the test only two cases fail, they all should fail.

Comment (kingthorin) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: For your first bullet can you be more specific? Or provide some examples?

Comment (thc202) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: e.g. `\u A`, `\x B`, `/* not closed …`.

Comment (kingthorin) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: Got all that, I think.

Comment (thc202) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: - Missing cases for `\u{HHH}` (1 to 6 digits long) and octal, unicode and hex broken half way through, from looking at the latest push.
- The hex escape is just two chars (the other ones are not really needed).
- There should also be a case for single quote strings and template literals (containing comment chars to make sure those are properly accepted).

Comment (kingthorin) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: Updated, I'm not sure what you wanted me to change for the second bullet.

Comment (thc202) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: `\xHH` not `\xHHHH` the latter is just one hex escape followed by two chars (case `w/ malformed (mid) hex escape` is incorrect, to be middle it would be `\xH H`).

Comment (kingthorin) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: Thanks, tweaked.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java [hunk 4]
```diff
@@ -476,11 +482,157 @@ private static RedirectType isRedirected(String payload, HttpMessage msg) {
     }
 
     private static boolean isRedirectPresent(Pattern pattern, String value) {
-        Matcher matcher = pattern.matcher(value);
+        // We know the payload is present.
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: I think this can go, given the explicit payload check being done before.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java [hunk 5]
```diff
@@ -476,11 +482,157 @@ private static RedirectType isRedirected(String payload, HttpMessage msg) {
...
+        // Ensure the value has something we're likely interested in before dealing with comments
+        if (JS_PRE_CHECKS.stream().noneMatch(chk -> StringUtils.containsIgnoreCase(value, chk))) {
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: I'd suggest keeping the previous check (even if that means matching twice), instead of adding additional checks which must be kept in sync with the patterns.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java [hunk 6]
```diff
@@ -476,11 +482,157 @@ private static RedirectType isRedirected(String payload, HttpMessage msg) {
...
+        for (String comment : extractedComments) {
+            valueWithoutComments = valueWithoutComments.replace(comment, "");
+        }
+
+        Matcher matcher = pattern.matcher(valueWithoutComments);
         return matcher.find()
                 && StringUtils.startsWithIgnoreCase(matcher.group(1), HttpHeader.HTTP);
     }
 
+    enum State {
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: private

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java [hunk 7]
```diff
@@ -476,11 +482,157 @@ private static RedirectType isRedirected(String payload, HttpMessage msg) {
...
+                        state = State.SLASH;
+                    }
+                    break;
+            }
+        }
+
+        // Handle unterminated comments at EOF
+        if (state == State.LINE_COMMENT) {
+            comments.add(currentComment.toString());
+        } else if (state == State.BLOCK_COMMENT || state == State.BLOCK_COMMENT_STAR) {
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: This should be all the same if.

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java (hunk 3 above)
Comment (thc202) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: The parsing logic keeps changing everyday which is fine but that means the tests should be updated to cover the new cases, there are still conditions not covered, which should to make sure there are no bugs.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java (hunk 5 above)
Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: You mean the domain check? Then drop the simple JS checks?

Comment (thc202) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: The `pattern` check. Yes, the `JS_PRE_CHECKS`.

Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: So do `matcher.find` both before and after?

Comment (thc202) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Right.

**Comment by kingthorin** [comment]
Got all those I believe

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java (hunk 3 above)
Comment (kingthorin) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: It was changing often to cover cases you kept thinking of, which is great because I wasn't seeing/catching them, and I thought I was extending the test inputs to cover things but maybe not sufficiently.

Just added a few for nested comments 🤷‍♂️

Comment (kingthorin) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: Found, added, and updated code for more edge cases.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java (hunk 5 above)
Comment (thc202) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Worth doing the `StringUtils.startsWithIgnoreCase` as well, better extract a method.

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java [hunk 8]
```diff
@@ -635,4 +681,210 @@ void shouldFindLocationUrl(String input) {
...
+        }
+
+        @ParameterizedTest(name = "{0}")
+        @MethodSource("commentProvider")
+        void shouldFindAComment(String name, String content) {
+            // Given / When
+            Set<String> comments =
+                    ExternalRedirectScanRule.extractJsComments(BODY_TEMPLATE.formatted(content));
+            // Then
+            assertThat(comments, hasSize(1));
```
Comment (thc202) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: The numbers are not enough, it should be checked the content as well otherwise we don't know if the boundaries are properly detected.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java (hunk 5 above)
Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Done & done

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java (hunk 8 above)
Comment (kingthorin) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: I've adjusted this. Indeed some of the results are perhaps not as we'd wish. I've added comments to the situations I feel are debatable .

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java [hunk 9]
```diff
@@ -635,4 +681,260 @@ void shouldFindLocationUrl(String input) {
...
+                            "/*  comment \\n // example */"),
+                    Arguments.of(
+                            "Inline block",
+                            "console.log(\"example\"); /* console.log('comment'); */",
+                            "/* console.log('comment'); */"),
+                    Arguments.of(
+                            "Inline incomplete block",
+                            // What do we really want in this case?
+                            "console.log(\"example\"); /* console.log('comment'); ",
+                            "/* console.log('comment'); \n</script></head><body><H1>Redirect</H1></body></html>"),
```
Comment (kingthorin) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: For this rule I think we only deal with JS inside HTML, so capturing to the closing script tag is probably reasonable.

In the future we could have a boolean that indicates if the end of script tag is used, or some other behavior to pure JS files.

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java [hunk 10]
```diff
@@ -635,4 +681,260 @@ void shouldFindLocationUrl(String input) {
...
+                    Arguments.of(
+                            "Division before comment",
+                            "var x = 42 / 2; // division is not regex",
+                            "// division is not regex"),
+                    Arguments.of(
+                            "Template expression with inner comment",
+                            // Is this the behavior we want? I believe this should either be both or
+                            // just trailing. The current result seems wrong to me.
+                            "console.log(`outer ${ /* inner comment */ 42 }`); // trailing comment",
+                            "/* inner comment */"),
```
Comment (kingthorin) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: I propose that this should simply capture the trailing comment.

Currently only the inner comment is captured.

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java [hunk 11]
```diff
@@ -635,4 +681,260 @@ void shouldFindLocationUrl(String input) {
...
+                    Arguments.of(
+                            "Nested template with string containing comment-like text",
+                            "console.log(`outer ${ 'string // not comment' }`); // real comment",
+                            "// real comment"),
+                    // This results in a zero comments situation. I believe that's likely fine as
+                    // the unterminated literal should be an app failure on it's own
+                    //                    Arguments.of(
+                    //                            "Unterminated template literal",
+                    //                            "console.log(`unterminated template ${1+1} //
+                    // comment not terminated", "?"),
```
Comment (kingthorin) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: Thoughts?

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java (hunk 9 above)
Comment (thc202) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: I'm good with that.
In the future we should use a proper battle tested parser.

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java (hunk 11 above)
Comment (thc202) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: What browsers do it's how it should behave.

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java (hunk 9 above)
Comment (kingthorin) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: I would be happy to use a parser, but the only options I could see seemed to have a lot of overhead, used internal APIs subject to change, or just ignored comments 😞 

If you have a suggestion I'm open to it.

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java (hunk 10 above)
Comment (thc202) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: Ideally it should capture both as to not lead to FPs (even if unlikely) as is the purpose of these changes.

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java (hunk 11 above)
Comment (kingthorin) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: So I tried this in Firefox, dev tools console:
```js
const a = 5;
const b = 10;
console.log(`Fifteen is ${a + b} and
not ${2 * a + b}.`);
```

It prints as expected.  So I tried (remove trailing backtick):
```js
const a = 5;
const b = 10;
console.log(`Fifteen is ${a + b} and
not ${2 * a + b}.);
```

It refused to let me execute it.
So I tried it on JS fiddle, which resulted in: `"[Line 8] SyntaxError: `` literal not terminated before end of script"`
Chrome dev tools console results in: `Uncaught SyntaxError: Unexpected end of input`

Comment (thc202) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: Preferable to do that in a script block.

Comment (kingthorin) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRuleUnitTest.java]: With, ex:
```html
<html>
<body>
<H1>test</h1>
<script>
const a = 5;
const b = 10;
console.log(`Fifteen is ${a + b} and
not ${2 * a + b}.);
</script>
</body>
</html>
```

The errors are similar, different line references or slightly different text, but still failures.

JSFiddle: `"[Line 1] SyntaxError: `` literal not terminated before end of script"`
Chrome: `Uncaught SyntaxError: Unexpected end of input (at tmpl_test.html:9:1)`
Safari: 
image

**Comment by kingthorin** [comment]
I think this is ready for review now. I believe I've addressed everything discussed.

**Comment by kingthorin** [comment]
Maybe this is a preferable solution (using Rhino), it handles all the existing test cases added here except "Inline incomplete block", which is kind of understandable. I doubt that browsers would be able to handle JavaScript in that case either.

https://github.com/zaproxy/zap-extensions/compare/main...kingthorin:zap-extensions:redir-wavsep-fps-rhino?expand=1

Let me know what you think, not a full review just a gutt feel: "Ya that's better" "naw let's stick with the state machine" kinda thing.

**Comment by thc202** [comment]
As said before any proven parser is preferable to manual parsing, the question is more which one. I don't have a preference as long it's being actively maintained (though using the engine we already bundle by default would be even better but that assumes it's possible).

**Comment by kingthorin** [comment]
Okay I've switched this over. Mainly I wanted to make sure that people were okay with using Rhino. It's maintained, its JS feature support is a bit behind but for the purpose here it should be totally fine.

**Comment by kingthorin** [comment]
Checking the failure.

**Comment by kingthorin** [comment]
Fixed

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java (hunk 5 above)
Comment (thc202) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Maybe this part was lost with the other changes but the `startsWithIgnoreCase` and method extraction was not done.

Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Done

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java [hunk 12]
```diff
@@ -478,10 +484,44 @@ private static RedirectType isRedirected(String payload, HttpMessage msg) {
...
+            try {
+                CompilerEnvirons env = new CompilerEnvirons();
+                env.setRecordingComments(true);
+                Parser parser = new Parser(env, env.getErrorReporter());
+                AstRoot ast = parser.parse(elem.getContent().toString(), null, 1);
+                if (ast.getComments() != null) {
+                    ast.getComments().forEach(comment -> comments.add(comment.getValue()));
+                }
+            } catch (EvaluatorException ee) {
+                LOGGER.debug(ee.getMessage());
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Nit, add a stat? Would be good to know how many parsing errors are happening.

**Comment by thc202** [comment]
Thank you!

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java (hunk 12 above)
Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Added

Comment (thc202) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Use the ID instead? That's what we use in the other ascan rule stats.

Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Will tweak in a bit

Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Adjusted.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java [hunk 13]
```diff
@@ -478,10 +485,42 @@ private static RedirectType isRedirected(String payload, HttpMessage msg) {
...
 
+    /** Visibility increased for unit testing purposes only */
+    protected static Set<String> extractJsComments(String jsSource) {
+        Set<String> comments = new HashSet<>();
+        try {
+            CompilerEnvirons env = new CompilerEnvirons();
+            env.setRecordingComments(true);
+            Parser parser = new Parser(env, env.getErrorReporter());
+            // Rhino drops a character when the snippet ends with a single line comment so add a newline
+            AstRoot ast = parser.parse(jsSource + "\n", null, 1);
```
Comment (kingthorin) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: This can be seen by leveraging `debugPrint()` on the ast or comment, the length of the comment is always reported one character short 😞

Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Reported upstream: https://github.com/mozilla/rhino/issues/2151

Comment (kingthorin) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java]: Without a doubt when others read this on Monday you'll wonder why this wasn't relevant before. But you can see from the simple example in the linked issue that it is a Rhino issue. I'm not sure, I didn't get to dig into that yet. 

I can but I don't want it to derail this. Let me know if you prefer that I roll back to the previous state.
